drivers/audio/rk3xi2sbus/buspdo.cpp

```diff
@@ -0,0 +1,328 @@
+        Header);
+    rhs = CONTAINING_RECORD(SecondIdentificationDescription,
+        PDO_IDENTIFICATION_DESCRIPTION,
+        Header);
+
+    RKI2SBusPrint(DEBUG_LEVEL_INFO, DBG_INIT,
+        "%s\n", __func__);
+
+    return (lhs->FdoContext == rhs->FdoContext) &&
+        (RtlCompareMemory(&lhs->CodecIds, &rhs->CodecIds, sizeof(lhs->CodecIds)) == sizeof(lhs->CodecIds));
+}
+
+VOID
+Bus_EvtChildListIdentificationDescriptionCleanup(
+    _In_ WDFCHILDLIST DeviceList,
+    _Inout_ PWDF_CHILD_IDENTIFICATION_DESCRIPTION_HEADER IdentificationDescription
+)
+/*++
+Routine Description:
+    It is called to free up any memory resources allocated as part of the description.
+    This happens when a child device is unplugged or ejected from the bus.
+    Memory for the description itself will be freed by the framework.
+Arguments:
+    DeviceList - Handle to the default WDFCHILDLIST created by the framework.
+    IdentificationDescription - Description of the child being deleted
+Return Value:
+--*/
+{
+    UNREFERENCED_PARAMETER(DeviceList);
+    UNREFERENCED_PARAMETER(IdentificationDescription);
+}
+
+NTSTATUS
+Bus_EvtDeviceListCreatePdo(
+    WDFCHILDLIST DeviceList,
+    PWDF_CHILD_IDENTIFICATION_DESCRIPTION_HEADER IdentificationDescription,
+    PWDFDEVICE_INIT ChildInit
+)
+/*++
+Routine Description:
+    Called by the framework in response to Query-Device relation when
+    a new PDO for a child device needs to be created.
+Arguments:
+    DeviceList - Handle to the default WDFCHILDLIST created by the framework as part
+                        of FDO.
+    IdentificationDescription - Decription of the new child device.
+    ChildInit - It's a opaque structure used in collecting device settings
+                    and passed in as a parameter to CreateDevice.
+Return Value:
+    NT Status code.
+--*/
+{
+    PPDO_IDENTIFICATION_DESCRIPTION pDesc;
+
+    PAGED_CODE();
+
+    pDesc = CONTAINING_RECORD(IdentificationDescription,
+        PDO_IDENTIFICATION_DESCRIPTION,
+        Header);
+
+    RKI2SBusPrint(DEBUG_LEVEL_INFO, DBG_INIT,
+        "%s\n", __func__);
+
+    return Bus_CreatePdo(WdfChildListGetDevice(DeviceList),
+        ChildInit,
+        pDesc);
+}
+
+NTSTATUS
+Bus_CreatePdo(
+    _In_ WDFDEVICE       Device,
+    _In_ PWDFDEVICE_INIT DeviceInit,
+    _In_ PPDO_IDENTIFICATION_DESCRIPTION           Desc
+)
+{
+    UNREFERENCED_PARAMETER(Device);
+
+    NTSTATUS                    status;
+    PPDO_DEVICE_DATA            pdoData = NULL;
+    WDFDEVICE                   hChild = NULL;
+    WDF_QUERY_INTERFACE_CONFIG  qiConfig;
+    WDF_OBJECT_ATTRIBUTES       pdoAttributes;
+    WDF_DEVICE_PNP_CAPABILITIES pnpCaps;
+    WDF_DEVICE_POWER_CAPABILITIES powerCaps;
+    DECLARE_CONST_UNICODE_STRING(deviceLocation, L"Rockchip I2S Controller");
+    DECLARE_UNICODE_STRING_SIZE(buffer, MAX_INSTANCE_ID_LEN);
+    DECLARE_UNICODE_STRING_SIZE(deviceId, MAX_INSTANCE_ID_LEN);
+    DECLARE_UNICODE_STRING_SIZE(compatId, MAX_INSTANCE_ID_LEN);
```

Comment:
Up to you, but I would have just defined a `DECLARE_CONST_UNICODE_STRING(strRockchip_I2S_Controller, "L"Rockchip I2S Controller")` for each of the constant strings.

If the function wants a non-const*, you can usually just cast, `const_cast<UNICODE_STRING*>(&strRockchip_I2S_Controller)`.

Reply:
We may use these as non-static members in the future depending on how the driver is re-used for SPDIF / DisplayPort audio. This driver code was originally from an HD Audio bus driver (that got stripped down), but it may be used to enumerate SPDIF as well as I2S in the near-future